heareval/tasks/speech_commands.py

```diff
@@ -175,8 +175,10 @@ def get_metadata_attrs(self, process_metadata):
             # In this case we take the slug and remove the -nohash- as described
             # This nohash removal allows for the speech of similar person to be in the
             # same dataset. Such type of data specific requirements might be there.
+            # [what does this mean??]
             # in the readme of google speech commands. we want to keep similar people
             # in the same group - test or train or val
+            # [so do we do this or not?]
```

Comment:
Can someone explain what this is about? @jorshi @khumairraj ? I think we should talk this over.

Reply:
This maybe happens automatically with the hashing? But not sure -- @khumairraj ?

Reply:
https://github.com/neuralaudio/hear2021-eval-kit/issues/93

Reply:
`.apply(lambda relpath: re.sub(r"-nohash-.*$", "", relpath))` This line removes the number and makes sure the hash of an audio file corresponds to the user id who recorded the example.
For example - `a_1, a_2, b_1, b_2`. If we hash the full name, the hash would not guarantee that the data by `a` and by `b` remains in different splits. This is a split condition to make sure the speech examples by the same person remains in the same split. Such a type of data specific split determination requires that we produce the hash in the metadata as it depends on the dataset. Such a requirement might not be there in another dataset and we might need to sample randomly(as is the case with others).

Reply:
But since we are no more relying on the hash to decide the split(we are using the text split file with the examples), and rather just using the hash for subsampling, we can remove this. But still, this ensures, when we subsample, we subsample people recording the voice and not the data points. We can decide on one way or the other and remove it accrodingly.
`process_metadata = process_metadata.sort_values(by="filename_hash").iloc[:max_files]`

Reply:
@khumairraj Yes, I understand. I still think this is written in a confusing way and it mixes several concerns.

1) It's hard to understand from looking at the code what it is doing, and it would be nice to clean up the comments and make things more obvious.
2) The hash is potentially used for several things: a) Splitting into train / validation in a deterministic way. b) Subsampling a split in a determinstic way. Right now, with hash = person, b is broken because the behavior of instances for a particular person is not deterministic. We really need to think through these abstractions so that the semantics are obvious and clear.

The simplest thing might be have a separate split_hash and subsample_hash, and do splitting in Luigi for anything that doesn't have a validation set.